### ISSUE TITLE
Rotary encoder works only on every second step

### ISSUE THREAD

**Issue by tomek-szczesny** [hardkernel/smartpower3#4]
The main knob is really frustrating, because it's hard to predict the outcome of another rotation step. 
I believe this is a bug in firmware rather than hardware issue. After all, the direction is always interpreted correctly.

**Comment by leeseungcheol** [comment]
I have recognized the issues.

But If the rotary has a single step, we get more unpredictable results. Please share any ideas whenever you have.

Thank you.

**Comment by tomek-szczesny** [comment]
Well, I'd have to dig into the code, in this case. But it's good to know there was a reason for it, and solving it will be more challenging.

Now when polling is implemented instead of interrupts for buttons, it may be reused to handle encoder instead of that pre-made library that fails to do its job.

**Comment by cedel1** [comment]
Hmm, I have looked at the code and played with the settings. You may want to try https://github.com/hardkernel/smartpower3/pull/26 and maybe play with the settings there.

I have come to the following discoveries:
- The encoder library seems to work reasonably fine. If set to full quad mode, every step is registered. Increasing the filter value also helps.
- The main problem (at least from the encoder usage point) is that the task that reads encoder steps (counting) and passes the value to the rest of the code has quite a large tick blocking value set - 100. Compare that with the screen blocking time (10). Combine that with values in "dial accelerator" in `countEncoder` method.
- What happens is that while the counts are probably correct, screen refreshes faster than data comes from the encoder task - and you get "tearing", disconnection from what you see on the screen and your expectation of what should happen.
- The perceived effect is different depending on which setting is changed - compare rotating the encoder fast while selecting which A/V value to change and actually changing the individual value setting.

Possible ways to get more usable and pleasing encoder response:
- Change the setting to full quad to get count at every encoder "click".
- Decrease the encoder task tick block value. Setting that value very low (below the screen refresh rate) makes the encoder behave rather nicely, but that is impractical, as there are more important things the processor should do than serving encoder.
- Make the counting method ("dial accelerator") behave differently depending on what setting is currently being changed. Slow down when selecting A/V value to change and make the accelerator behave faster when actually changing the value (to make it faster to go from 3V to 20V - maybe even changing whole units when turning fast and decimal places in "slow mode")
- As the encoder is used occasionally (compared to other functions of the device), change task blocking time depending on if the encoder is currently in use or not or depending on what setting is being changed.


A reasonable compromise of all the possible solutions has to be found. Unfortunately, I currently don't have the means and knowledge to measure the impact of lowering the blocking time exactly and reliably. After all, having a good user interface is important, but it must not adversely affect the main functions of the device - mainly controlling the output and logging.

**Comment by cedel1** [comment]
BTW, the hardest SP3 setting to test this that I have found is to set the highest console speed (921600 bps) and the shortest sampling time (5ms), UDP port set and WiFi set. Then try to rotate the encoder fast (slow step seems to work OK). This setting seems to tax the settings and system quite a lot. Lowering the sampling speed to 10ms improves the behavior.

Also, I've just checked with current master branch build and if I turn the encoder really fast, then it also jumps on V/A selection.
